src/pc/common/PaletteContainer.js

```diff
@@ -22,12 +23,15 @@ class DirectPaletteContainer {
 
   write (smartBuffer) {
     smartBuffer.writeUInt8(this.data.bitsPerValue)
-    varInt.write(smartBuffer, this.data.length())
+    if (!this.noSizePrefix) varInt.write(smartBuffer, this.data.length())
     this.data.writeBuffer(smartBuffer)
   }
 
-  readBuffer (smartBuffer) {
-    this.data.readBuffer(smartBuffer, varInt.read(smartBuffer) * 2)
+  readBuffer (smartBuffer, bitsPerValue) {
+    const longs = this.noSizePrefix
+      ? Math.ceil(this.data.capacity / Math.floor(64 / bitsPerValue))
```

Comment:
Previously there was a bug here in that it used a constant instead of checking underlying storage capacity:
> The core issue is that your generic IndirectPaletteContainer and DirectPaletteContainer classes are hardcoded to calculate their size based on a block section's volume (16x16x16 = 4096), even when they are being used for a biome section (4x4x4 = 64).
> The solution is to make the readBuffer methods in DirectPaletteContainer and IndirectPaletteContainer use their own capacity instead of the hardcoded BLOCK_SECTION_VOLUME. This will correctly calculate 1 long for biomes and 64 longs for blocks (when bitsPerValue is 1).

Biomes are no longer per block, but per chunks of a chunk column